### PR TITLE
Python: Pass kernel args to tool calls

### LINKED ISSUE #5454: Python: unable to get kernel arguments (chat history) in a custom kernel function while using auto tool calls
**Describe the bug**
While using auto tool calls, the only arguments that the custom kernel function can use are defined by the LLM (mainly from the tool call object that is created; however the types may not match). If I configure a custom kernel function (from a method) to accept a string argument and a KernelArguments type arguments, the arguments will be of type string, all while missing the actual kernel arguments configured before invoking the function.

**Expected Behavior**
I should be able to make use of the kernel arguments, and things that are contained in it, like the chat_history object inside my kernel function.

**Platform**
 - Language: Python
 - Source: Python 0.9.1b1

### PR DESCRIPTION
**Pull request by moonbox3**
### Motivation and Context

After introducing auto tool calls, they mostly work well for primitive type arguments that are handled by the LLM; however, it may be that the user would like to include other kernel arguments to be used as part of their custom functions, that are of type like KernelArguments.

### Description

This PR combines the kernel arguments plus the required arguments for tool calls. It also includes the kernel function's metadata, which has the kernel arguments while storing the chat message after performing tool calling. 
- This closes #5454 

### Contribution Checklist

- [X] The code builds clean without any errors or warnings
- [X] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [X] All unit tests pass, and I have added new tests where possible
- [X] I didn't break anyone :smile:

### CODE REVIEW

File: python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py [hunk 1]
```diff
@@ -329,17 +345,18 @@ async def _process_tool_calls(
         result: Union[OpenAIChatMessageContent, OpenAIStreamingChatMessageContent],
         kernel: "Kernel",
         chat_history: ChatHistory,
+        arguments: "KernelArguments",
     ) -> None:
         """Processes the tool calls in the result and return it as part of the chat history."""
         logger.info(f"processing {len(result.tool_calls)} tool calls")
+        args_cloned = copy(arguments)
```
Comment (juliomenendez) [inline python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py:352]: Nothing major but I believe the `.copy` method for dict/list is faster than the `copy` module.

**Review by juliomenendez — CHANGES_REQUESTED** [review]

File: python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py [hunk 2]
```diff
@@ -66,15 +68,20 @@ async def complete_chat(
         """
         tool_call_behavior = self._get_tool_call_behavior(settings)
         kernel = kwargs.get("kernel", None)
-        if tool_call_behavior.auto_invoke_kernel_functions and kernel is None:
-            raise ServiceInvalidExecutionSettingsError("The kernel argument is required for OpenAI tool calling.")
+        arguments = kwargs.get("arguments", None)
+        if tool_call_behavior.auto_invoke_kernel_functions and kernel and arguments is None:
```
[the remark this answers is not in the document]
Comment (moonbox3) [reply python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py:72]: I wanted if all are none, but we can move it to your suggestion.

[the remark this answers is not in the document]
Comment (moonbox3) [reply python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py:72]: Updated

[the remark this answers is not in the document]
Comment (juliomenendez) [reply python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py:72]: If all are none wouldn't catch the case where kernel or arguments are none, right?

[the remark this answers is not in the document]
Comment (moonbox3) [reply python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py:72]: Yeah good point, thanks.

**Comment by markwallace-microsoft** [comment]
<a href="https://github.com/microsoft/semantic-kernel/blob/4d9abec85681ece38fb879a2fb24dc5f081f2638/README.md">Py3.8 Test Coverage</a><details><summary>Python 3.8 Test Coverage Report • </summary><table><tr><th>File</th><th>Stmts</th><th>Miss</th><th>Cover</th><th>Missing</th></tr><tbody><tr><td colspan="5"><b>semantic_kernel/connectors/ai/open_ai/services</b></td></tr><tr><td>&nbsp; &nbsp;<a href="https://github.com/microsoft/semantic-kernel/blob/4d9abec85681ece38fb879a2fb24dc5f081f2638/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py">open_ai_chat_completion_base.py</a></td><td>159</td><td>62</td><td>61%</td><td><a href="https://github.com/microsoft/semantic-kernel/blob/4d9abec85681ece38fb879a2fb24dc5f081f2638/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py#L100">100</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/4d9abec85681ece38fb879a2fb24dc5f081f2638/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py#L116-L119">116&ndash;119</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/4d9abec85681ece38fb879a2fb24dc5f081f2638/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py#L121-L124">121&ndash;124</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/4d9abec85681ece38fb879a2fb24dc5f081f2638/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py#L141-L144">141&ndash;144</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/4d9abec85681ece38fb879a2fb24dc5f081f2638/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py#L153-L156">153&ndash;156</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/4d9abec85681ece38fb879a2fb24dc5f081f2638/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py#L166-L189">166&ndash;189</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/4d9abec85681ece38fb879a2fb24dc5f081f2638/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py#L198-L200">198&ndash;200</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/4d9abec85681ece38fb879a2fb24dc5f081f2638/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py#L217-L219">217&ndash;219</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/4d9abec85681ece38fb879a2fb24dc5f081f2638/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py#L242">242</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/4d9abec85681ece38fb879a2fb24dc5f081f2638/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py#L250">250</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/4d9abec85681ece38fb879a2fb24dc5f081f2638/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py#L256-L262">256&ndash;262</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/4d9abec85681ece38fb879a2fb24dc5f081f2638/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py#L273-L279">273&ndash;279</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/4d9abec85681ece38fb879a2fb24dc5f081f2638/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py#L315-L321">315&ndash;321</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/4d9abec85681ece38fb879a2fb24dc5f081f2638/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py#L337">337</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/4d9abec85681ece38fb879a2fb24dc5f081f2638/python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py#L343-L345">343&ndash;345</a></td></tr><tr><td colspan="5"><b>semantic_kernel/functions</b></td></tr><tr><td>&nbsp; &nbsp;<a href="https://github.com/microsoft/semantic-kernel/blob/4d9abec85681ece38fb879a2fb24dc5f081f2638/python/semantic_kernel/functions/kernel_function.py">kernel_function.py</a></td><td>57</td><td>2</td><td>96%</td><td><a href="https://github.com/microsoft/semantic-kernel/blob/4d9abec85681ece38fb879a2fb24dc5f081f2638/python/semantic_kernel/functions/kernel_function.py#L119">119</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/4d9abec85681ece38fb879a2fb24dc5f081f2638/python/semantic_kernel/functions/kernel_function.py#L138">138</a></td></tr><tr><td>&nbsp; &nbsp;<a href="https://github.com/microsoft/semantic-kernel/blob/4d9abec85681ece38fb879a2fb24dc5f081f2638/python/semantic_kernel/functions/kernel_function_from_prompt.py">kernel_function_from_prompt.py</a></td><td>123</td><td>3</td><td>98%</td><td><a href="https://github.com/microsoft/semantic-kernel/blob/4d9abec85681ece38fb879a2fb24dc5f081f2638/python/semantic_kernel/functions/kernel_function_from_prompt.py#L164">164</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/4d9abec85681ece38fb879a2fb24dc5f081f2638/python/semantic_kernel/functions/kernel_function_from_prompt.py#L191">191</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/4d9abec85681ece38fb879a2fb24dc5f081f2638/python/semantic_kernel/functions/kernel_function_from_prompt.py#L287">287</a></td></tr><tr><td colspan="5"><b>semantic_kernel/planners/function_calling_stepwise_planner</b></td></tr><tr><td>&nbsp; &nbsp;<a href="https://github.com/microsoft/semantic-kernel/blob/4d9abec85681ece38fb879a2fb24dc5f081f2638/python/semantic_kernel/planners/function_calling_stepwise_planner/function_calling_stepwise_planner.py">function_calling_stepwise_planner.py</a></td><td>92</td><td>33</td><td>64%</td><td><a href="https://github.com/microsoft/semantic-kernel/blob/4d9abec85681ece38fb879a2fb24dc5f081f2638/python/semantic_kernel/planners/function_calling_stepwise_planner/function_calling_stepwise_planner.py#L115-L166">115&ndash;166</a>, <a href="https://github.com/microsoft/semantic-kernel/blob/4d9abec85681ece38fb879a2fb24dc5f081f2638/python/semantic_kernel/planners/function_calling_stepwise_planner/function_calling_stepwise_planner.py#L197-L202">197&ndash;202</a></td></tr><tr><td><b>TOTAL</b></td><td><b>5044</b></td><td><b>990</b></td><td><b>80%</b></td><td>&nbsp;</td></tr></tbody></table></details>

## Python 3.8 Unit Test Overview
| Tests | Skipped | Failures | Errors | Time |
| ----- | ------- | -------- | -------- | ------------------ |
| 1059 | 11 :zzz: | 0 :x: | 0 :fire: | 23.140s :stopwatch: |

File: python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py [hunk 3]
```diff
@@ -329,17 +345,18 @@ async def _process_tool_calls(
...
+        args_cloned = copy(arguments)
         for tool_call in result.tool_calls:
             if tool_call.function is None:
                 continue
             logger.info(f"Calling {tool_call.function.name} function with args: {tool_call.function.arguments}")
             try:
-                func_result = await kernel.invoke(
-                    **tool_call.function.split_name_dict(), arguments=tool_call.function.to_kernel_arguments()
-                )
+                args_cloned.update(tool_call.function.parse_arguments())
```
Comment (eavanvalkenburg) [inline python/semantic_kernel/connectors/ai/open_ai/services/open_ai_chat_completion_base.py:358]: might be good to have parse_arguments in a seperate try except block, or to add a specific except for the one returned by parse_args, with auto invoke that can then be fed back to the model.

**Review by eavanvalkenburg — CHANGES_REQUESTED** [review]
small note, but overall looks good!

**Review by juliomenendez — APPROVED** [review]

**Review by eavanvalkenburg — APPROVED** [review]
Love it!

**Review by eavanvalkenburg — APPROVED** [review]

**Pull request merged**